### ISSUE TITLE
Incompatibility with GHC 9.*

### ISSUE THREAD

**Issue by Dessix** [airalab/hs-web3#125]
I'm not sure how to fix this or I'd submit a PR, but on GHC 9.2.1 via stackage `resolver: nightly-2022-02-01` I hit the below log when compiling, but not on 8.x:

```
web3-solidity                    > [14 of 17] Compiling Data.Solidity.Prim.Tuple
web3-solidity                    >            
web3-solidity                    > /tmp/stack-7f245c5571e282ad/web3-solidity-1.0.0.0/src/Data/Solidity/Prim/Tuple.hs:30:10: error:
web3-solidity                    >     • Illegal instance declaration for ‘Generic (OneTuple a)’
web3-solidity                    >         (All instance types must be of the form (T t1 ... tn)
web3-solidity                    >          where T is not a synonym.
web3-solidity                    >          Use TypeSynonymInstances if you want to disable this.)
web3-solidity                    >     • In the instance declaration for ‘Generic (OneTuple a)’
web3-solidity                    >    |       
web3-solidity                    > 30 | instance Generic (OneTuple a)
web3-solidity                    >    |          ^^^^^^^^^^^^^^^^^^^^
web3-solidity                    >            
web3-solidity                    > /tmp/stack-7f245c5571e282ad/web3-solidity-1.0.0.0/src/Data/Solidity/Prim/Tuple.hs:32:23: error:
web3-solidity                    >     • Illegal instance declaration for ‘AbiType (OneTuple a)’
web3-solidity                    >         (All instance types must be of the form (T t1 ... tn)
web3-solidity                    >          where T is not a synonym.
web3-solidity                    >          Use TypeSynonymInstances if you want to disable this.)
web3-solidity                    >     • In the instance declaration for ‘AbiType (OneTuple a)’
web3-solidity                    >    |       
web3-solidity                    > 32 | instance AbiType a => AbiType (OneTuple a) where
web3-solidity                    >    |                       ^^^^^^^^^^^^^^^^^^^^
web3-solidity                    >            
web3-solidity                    > /tmp/stack-7f245c5571e282ad/web3-solidity-1.0.0.0/src/Data/Solidity/Prim/Tuple.hs:35:22: error:
web3-solidity                    >     • Illegal instance declaration for ‘AbiGet (OneTuple a)’
web3-solidity                    >         (All instance types must be of the form (T t1 ... tn)
web3-solidity                    >          where T is not a synonym.
web3-solidity                    >          Use TypeSynonymInstances if you want to disable this.)
web3-solidity                    >     • In the instance declaration for ‘AbiGet (OneTuple a)’
web3-solidity                    >    |       
web3-solidity                    > 35 | instance AbiGet a => AbiGet (OneTuple a)
web3-solidity                    >    |                      ^^^^^^^^^^^^^^^^^^^
web3-solidity                    >            
web3-solidity                    > /tmp/stack-7f245c5571e282ad/web3-solidity-1.0.0.0/src/Data/Solidity/Prim/Tuple.hs:36:22: error:
web3-solidity                    >     • Illegal instance declaration for ‘AbiPut (OneTuple a)’
web3-solidity                    >         (All instance types must be of the form (T t1 ... tn)
web3-solidity                    >          where T is not a synonym.
web3-solidity                    >          Use TypeSynonymInstances if you want to disable this.)
web3-solidity                    >     • In the instance declaration for ‘AbiPut (OneTuple a)’
web3-solidity                    >    |       
web3-solidity                    > 36 | instance AbiPut a => AbiPut (OneTuple a)
```

**Comment by Dessix** [comment]
Adjusted the title to `9.*` as `9.0.1` produces the above errors; 9.2.1 isn't even available on stackage somehow, but overriding with `compiler: ghc-9.2.1` in `stack.yaml` produces a lot of other issues- including revealing that the dependency `animalcase` hasn't been maintained since 2014.

**Comment by Dessix** [comment]
Having spent the last half hour trying to figure out what all is involved- this, it looks like a switch to `cabal-install` instead of stack would simplify versioning greatly, since this is library code, and snapshots tend to be an antipattern there.

Regardless, I've determined a few parts that need updated:
- `base` must include `4.16`
- `animalcase` dependency should be removed due to lack of maintenance since 2014 and further incompatibility- it has a single usage of `toCamelCase` in `jsonrpc`; `casing` appears to be a viable candidate for replacement, or the function could be hand-written private with a unit test due to limited usage
- `aeson` must be `>=2.0.3.0 && <2.1` to support 9.2.1
- `memory` range must include `0.16` for 9.2.1
- `bytestring` range must include `0.11` for 9.2.1
- `template-haskell` range must include `0.18` for 9.2.1
- `http-client` must include 0.7
- `wide-word` has 9.2.1 compatibility but it needs published to hackage: erikd/wide-word#66
- `microlens` is importing an older version than supports 9.2.1- it may not even support it at all yet

(This list is not exhaustive- solving the above may reveal more)
